peps/pep-0758.rst

```diff
@@ -0,0 +1,432 @@
+URL will be defined separately, for the sake of example, let's say a package
+``foo`` externally hosts wheel files on ```https://foo.example.com``
+<https://foo.example.com>`__ and that host becomes unreachable. When an
+installer tool tries to download and install the package ``foo`` wheel, the
+download step will fail. The installer would then be able to query PyPI to
+provide a useful error message to the end user:
+
+- The installer reads the ``Owner`` field from the ``EXTERNAL-HOSTING`` file
+  inside the ``.rim`` zip file.
+- The installer queries PyPI for the support URI for the organization
+  owner of the externally hosted wheel.
+- An informative error message would then be displayed, e.g.:
+
+   The externally hosted wheel file ``foo-....whl`` could not be
+   downloaded. Please contact support@foo.example.com for help. Do not report
+   this to the PyPI administrators.
+
+Dismounting wheels
+==================
+
+It is generally very easy to produce a ``.rim`` file from an existing ``.whl``
+file. This could be done efficiently by a :pep:`518` build backend with an
+additional command line option, or a separate tool which takes a ``.whl`` file
+as input and creates the associated ``.rim`` file. To complete the analogy,
+the act of turning a ``.whl`` into a ``.rim`` is called "dismounting".  The
+steps such a tool would take are:
+
+- Accept as input the source ``.whl`` file, the organization owner of the
+  package, and URL at which the ``.whl`` will be hosted, and the support URI
+  to report download problems from. These could in fact be captured in the
+  ``pyproject.toml`` file, but that specification is out of scope for this
+  PEP.
+- Unzip the ``.whl`` and create the ``.rim`` zip archive.
+- Omit from the ``.rim`` file any path in the ``.whl`` that **isn't** rooted
+  at the ``.dist-info`` directory.
+- Calculate the hash of the source ``.whl`` file.
+- Add to the ``.rim`` archive the ``EXTERNAL-HOSTING`` file containing
+  the headers and their values as described above.
+
+Changes to tools
+================
+
+Theoretically, installer tools shouldn't need any changes, since when they
+have identified the wheel to download and install, they simply consult the
+download URLs returned by PyPI's Simple API. In practice though, tools such as
+``pip`` and ``uv`` may have constrained lists of hosts they will allow
+downloads from, such as PyPI's own ``pythonhosted.org`` domain.
+
+In this case, such tools will need to relax those constraints, but the exact
+policy for this is left to the installer tools themselves. Any number of
+approaches could be implemented, such as downloading the ``.rim`` file and
+verifying the ``EXTERNAL-HOSTING`` metadata, or simply trusting the external
+downloads for any wheel with a matching checksum.  They could also query PyPI
+for the project's organization owner, then verify that it has a support URI
+before trusting the download. They could warn the user when externally hosted
```

Comment:
Isn't this extant by construction? IOW I don't think this is a check that needs to be made?

Reply:
If I understand the question correctly, I think in general an installer wouldn't need to download the `.rim` file since all the information it needs is in the package metadata and download urls.  The only other thing that `E-H` file contains is the owner information, which AFAICT is not available in the [PEP 691 project detail JSON](https://peps.python.org/pep-0691/#project-detail).  Maybe it should be, but that's a different question.

Reply:
Sorry, I realize this may have been a bit opaque. I was responding to the installer verifying the project's organization has a support URI. Specifically:
> They could also query PyPI for the project's organization owner, then verify that it has a support URI
> before trusting the download.

We're already specifying that all organizations with projects using external hosting must have a contact URI, so I don't understand why pip would verify that if it is part of the contract pypi.org would provide.

FWIW I think it would be good to serve the EH file from PyPi. Since external hosting is a PyPi-centric design, I don't think we need to touch the simple API. Installers can just query for the EH file if they see that the file location is outside of `pythonhosted.org` when doing their normal simple API search.

Reply:
> We're already specifying that all organizations with projects using external hosting must have a contact URI, so I don't understand why pip would verify that if it is part of the contract pypi.org would provide.

Ah, makes sense.

> FWIW I think it would be good to serve the EH file from PyPi. Since external hosting is a PyPi-centric design, I don't think we need to touch the simple API. Installers can just query for the EH file if they see that the file location is outside of pythonhosted.org when doing their normal simple API search.

The EH file itself or the whole `.rim` file.  The latter will already be accessible, so I don't think PyPI needs to unpack the `.rim` and serve the EH file independently.